poc/vdaf_poplar1.sage

```diff
@@ -11,198 +11,206 @@ from sagelib.vdaf import Vdaf, test_vdaf
-        sketch_share = [a_share, b_share, c_share]
+        value = Poplar1.Idpf.eval(agg_id,
+                                  idpf_public_share,
+                                  key,
+                                  level,
+                                  prefixes)
+
+        # Leader expands randomness
+        if agg_id == 0:
+            prg = Poplar1.Idpf.Prg(corr_seed, dst + byte(255))
+            auth = prg.next_vec(Poplar1.Idpf.FieldInner,
+                                Poplar1.Idpf.BITS - 1)
+            auth += prg.next_vec(Poplar1.Idpf.FieldLeaf, 1)
+
+        # Prepare one-hotness and boundnedness verification
+        # and output share
+        # sum over all prefixes for boundedness verification
         out_share = []
-        for (i, r) in enumerate(verify_rand):
-            (data_share, auth_share) = value[i]
-            sketch_share[0] += data_share * r
-            sketch_share[1] += data_share * r^2
-            sketch_share[2] += auth_share * r
+        data_sum = Field(0)
+        auth_sum = Field(0)
+        corr_results = []
+        correction_word = corr_inner[level] if level < Poplar1.Idpf.BITS-1 \
+            else corr_leaf
+        for i in range(len(prefixes)):
+            (data_share, auth_share, indicator_share) = value[i]
+            z = Field(1) if agg_id == 0 else Field(-1)
+            corrected_x = Poplar1.hash1(level,
+                                        prefixes[i],
+                                        [z * data_share, z * auth_share])
+            result = Poplar1.correct(corrected_x,
+                                     indicator_share,
+                                     correction_word)
+            corr_results.append(result)
+            data_sum -= z * data_share
+            auth_sum -=  z * auth_share
             out_share.append(data_share)
 
-        prep_mem = sketch_share \
-                    + [A_share, B_share, Field(agg_id)] \
-                    + out_share
-        return (b'ready', level, prep_mem)
+        corr_check = Poplar1.hash3(corr_results, level)
+
+        bound_msg_part_1 = b''
+        bound_msg_part_2 = b''
+        if agg_id == 0:
+            offset = OS2IP(gen_rand(Poplar1.BOUND)) % (Poplar1.BOUND + 1)
+            for i in range(0, offset):
+                bound_msg_part_2 += Poplar1.hash2(Field(i) + data_sum,
+                                            Field(i) * auth[level] + auth_sum,
+                                            level)
+            for i in range(offset, Poplar1.BOUND+1):
+                bound_msg_part_1 += Poplar1.hash2(Field(i) + data_sum,
+                                            Field(i) * auth[level] + auth_sum,
+                                            level)
+        bound_msg = bound_msg_part_1 + bound_msg_part_2 if agg_id == 0 else b''
+
+        check_bound = b'' if agg_id == 0 \
+                    else Poplar1.hash2(data_sum, auth_sum, level)
```

Comment:
Doesn't `hash2()` need to be a random hash function unknown to the attacker? I.e., don't we want to pass in something dervied from `verify_key` here?

Reply:
I don't think it does? Hash_2 needs to be a random oracle for Privacy, in which case the attacker is Bob who knows verify_key (and the hash isn't private anyway), and for soundness, where the attacker genuinely has no reason to know H, we only require collision resistance.